modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1156,10 +1274,15 @@ void redeploy(IgniteUuid srvcId, ServiceConfiguration cfg,
 
         Collection<ServiceContextImpl> toInit = new ArrayList<>();
 
+        ServiceContextImpl firstInitialized = !metricsEnabled || ctxs.isEmpty() ? null : ctxs.iterator().next();
```

Comment:
I suggest keeping metric in a Map: service name -> method metrics. (if names are unique) Like `Map<String, Map<Method, HistogramMetricImpl>>`. WDYT?

Reply:
The map will be held by IgniteServiceProcessor. This would force using not-static LocalInvocationHandler. Also, there is the ServiceContext to keep something about certain service. Making the map outside it can make similar question: why not in service context?

Reply:
Several service instances can be deployed per node.  One service context per deployed instance. But one metric per node(all deployed instances of one service), isn't it?